board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c

```diff
@@ -87,6 +126,250 @@ int board_early_init_f(void)
+	 * should be same as the one programmed into QSPI flash
+	 */
+	if (size < 0x1000) {
+		printf("M4: Image size too small (%d)!\n", size);
+		return -EINVAL;
+	} else
+		printf("M4: Fit Image Size %d\n", size);
+
+	/* check FIT firmware tags */
+	if (!M4_FW_VALID(M4_WATERMARK(data))) {
+		printf("M4: Invalid image: tag=0x%x\n", M4_WATERMARK(data));
+		return -EINVAL;
+	}
+
+	len = round_down(M4_SIZE, mtd->erasesize);
+	if (spi_flash_erase(flash, 0, len))
+		printf("M4: Failed to erase spi flash\n");
+
+	len = round_up(size, mtd->writesize);
+	if (spi_flash_write(flash, 0, len, data)) {
+		printf("M4: Failed to write image to spi flash\n");
+		return -EIO;
+	}
+
+	optee_update_tee_hash(fit_hash, size);
+	printf("M4: Firmware upgraded from FIT\n");
+
+	return 0;
+}
+
+static int m4_upgrade_required(struct hash *fit_hash)
+{
+	uint8_t tee_hash[FIT_MAX_HASH_LEN];
+	int tee_hash_len = -1;
+
+	if (optee_get_tee_hash(tee_hash, &tee_hash_len, NULL))
+		return false;
+
+	if (tee_hash_len != fit_hash->len ||
+	    memcmp(tee_hash, fit_hash->value, tee_hash_len))
+		return true;
+
+	return false;
+}
+
+static int m4_check_hash(const void *fit, const int image_noffset,
+			 struct hash *fit_hash)
+{
+	int noffset = 0;
+	int i;
+
+	fdt_for_each_subnode(noffset, fit, image_noffset) {
+		const char *name = fit_get_name(fit, noffset, NULL);
+		if (!strncmp(name, FIT_HASH_NODENAME,
+			     strlen(FIT_HASH_NODENAME)))
+			goto process_m4_hash;
+	}
+
+	printf("M4: Can't get hash node\n");
+	return -EINVAL;
+
+process_m4_hash:
+
+	if (fit_image_hash_get_value(fit, noffset,
+				     &fit_hash->value, &fit_hash->len)) {
+		printf("M4: Can't get hash value property\n");
+		return -EINVAL;
+	}
+
+	printf("M4: FIT hash => ");
+	for (i = 0; i < fit_hash->len; i++)
+		printf("%x", fit_hash->value[i]);
+	printf("\n");
+
+	if (m4_upgrade_required(fit_hash) == false) {
+		printf("M4: upgrade not required\n");
+		return 0;
+	}
+
+	return 1;
+}
+
+static int m4_get_fit_node(const void *fit)
+{
+	int images_noffset;
+	const char *name;
+	int noffset;
+	int ndepth;
+	int count;
+
+	images_noffset = fdt_path_offset(fit, FIT_IMAGES_PATH);
+	if (images_noffset < 0) {
+		printf("M4: Can't find images parent node '%s' (%s)\n",
+		       FIT_IMAGES_PATH, fdt_strerror(images_noffset));
+		return -EINVAL;
+	}
+
+	for (ndepth = 0, count = 0,
+	     noffset = fdt_next_node(fit, images_noffset, &ndepth);
+			(noffset >= 0) && (ndepth > 0);
+			noffset = fdt_next_node(fit, noffset, &ndepth)) {
+
+		if (ndepth == 1) {
+			name = fit_get_name(fit, noffset, NULL);
+			if (!strncmp(name, "m4", 2))
+				return noffset;
+ 			count++;
+		}
+	}
+
+	printf("M4: Can't find the FPGA M4 node\n");
+
+	return -EINVAL;
+}
+
+static void m4_boot(struct spi_flash *flash)
+{
+	int hash_len = 0, tee_hash_len = 0, tee_payload_len = 0;
+	uint8_t tee_hash[FIT_MAX_HASH_LEN];
+	uint8_t hash[FIT_MAX_HASH_LEN];
+
+	if (spi_flash_read(flash, 0, M4_SIZE, (void *) M4_BASE)) {
```

Comment:
I think this could be changed to a memory clear (memset) statement using M4_SIZE.  (See below about shifting of `optee_get_tee_hash` to here (under the memset).

Then `spi_flash_read` using the `tee_payload_len` value for the size (rounded up).

This would avoid loading the entire SRAM space from QSPI which could be an attack vector as the later hash check only verifies the hash of up to `tee_payload_len`.